httpx/_client.py

```diff
@@ -682,7 +679,13 @@ def __init__(
             if proxy:
                 raise RuntimeError("Use either `proxy` or 'proxies', not both.")
 
-        allow_env_proxies = trust_env and app is None and transport is None
+        if app:
+            raise RuntimeError(
```

Comment:
I don't know what's deprecation policy but are you sure raise error here? isn't it breaking change for those using `app`?

I'd recommend keep docs, type-hints and use warning instead. then we can schedule the removal version and mention it in warning message (?)

Reply:
> I don't know what's deprecation policy but are you sure raise error here? isn't it breaking change for those using `app`?

It looks good to me if this breaking change is for 1.0. But do we want to keep these additional errors in 1.0?

Reply:
> It looks good to me if this breaking change is for 1.0. But do we want to keep these additional errors in 1.0?

So...  we'll need to have consistency around this.

Options might be...

* Include old argument styles for 1.0 with explicit errors, to aid migrations. Clean up later and drop the errors in a 1.1 release.
* Don't include old argument styles.

Reply:
At there, I'm more fan of what done for `proxies`:
https://github.com/encode/httpx/blob/419d3a9d80d0c4072f6cb58eeb306148ae89e2e9/httpx/_client.py#L676-L681

So, what I can suggest at here is:
1. deprecate and mention scheduled removal if any (e.g. "The 'proxies' argument is now deprecated and it will remove in v1.0.0")
2. minor/patch release
3. clear all deprecations in major release

Reply:
I think now is good time to discuss about deprecation policy, since there are few `1.0 proposal` PRs those trying to directly remove APIs without any deprecation.

Reply:
https://github.com/encode/httpx/pull/3069 would be a good place for us to discuss how we'd like to approach this.

Reply:
Via #3071 and since v1.0 is major upgrade, I'm now more interested to directly drop instead of deprecation.